tools/indexer/pkg/backfill/backfill.go

```diff
@@ -0,0 +1,171 @@
+
+	logger.Info("RecentMissing processed", "blocks", len(blocks))
+	return nil
+}
+
+// RecentBids backfills bid data for ALL recent slots (not just opted-in blocks)
+func recentBids(ctx context.Context, db *database.DB, httpc *retryablehttp.Client, relays []relay.Row, slots []int64) error {
+	logger := slog.With("component", "backfill")
+
+	for _, slot := range slots {
+		if ctx.Err() != nil {
+			break
+		}
+
+		for _, rr := range relays {
+			if ctx.Err() != nil {
+				break
+			}
+
+			fetchCtx, fcancel := context.WithTimeout(ctx, 5*time.Second)
+			bids, err := relay.FetchBuilderBlocksReceived(fetchCtx, httpc, rr.URL, slot)
+			fcancel()
+			if err != nil {
+				logger.Error("RecentBids fetch failed", "slot", slot, "relay_id", rr.ID, "relay_url", rr.URL, "error", err)
+				continue
+			}
+
+			rows := make([]database.BidRow, 0, len(bids))
+			for _, b := range bids {
+				if row, ok := relay.BuildBidInsert(slot, rr.ID, b); ok {
+					rows = append(rows, row)
+				}
+			}
+
+			if len(rows) > 0 {
+				insCtx, icancel := context.WithTimeout(ctx, 5*time.Second)
+				if err := db.InsertBidsBatch(insCtx, rows); err != nil {
+					icancel()
+					return fmt.Errorf("bids insert failed slot=%d relay_id=%d: %w", slot, rr.ID, err)
+				}
+				icancel()
+			}
+		}
+
+	}
+
+	logger.Info("RecentBids processed", "slots", len(slots))
+	return nil
+}
+
+// RunAll executes all backfill operations ensuring complete coverage
+func RunAll(ctx context.Context, db *database.DB, httpc *retryablehttp.Client, cfg *config.Config, relays []relay.Row) error {
+	logger := slog.With("component", "backfill")
+	logger.Info("Starting comprehensive backfill for ALL blocks (not just opted-in)")
+
+	// Channel to pass slot data from stage 1 to stages 2 & 3
+	slotChan := make(chan SlotData, cfg.BackfillBatch)
+	errCh := make(chan error, 1)
+
+	// Run recentMissing and collect slot data
+	go func() {
+		if err := recentMissing(ctx, db, httpc, cfg, cfg.BackfillLookback, cfg.BackfillBatch, slotChan); err != nil {
+			errCh <- err
+		}
+		close(slotChan)
+	}()
+
+	// Collect slots and validator data from channel
+	var slotsForBids []int64
+	var validatorsToCheck []SlotData
+
+	for data := range slotChan {
+		slotsForBids = append(slotsForBids, data.Slot)
+		if len(data.ValidatorPubkey) > 0 {
+			validatorsToCheck = append(validatorsToCheck, data)
+		}
+	}
+	select {
+	case err := <-errCh:
+		if err != nil {
+			logger.Error("RecentMissing failed", "error", err)
+			return err
+		}
+	default:
+	}
+	for _, v := range validatorsToCheck {
+		opted, err := ethereum.CallAreOptedInAtBlock(httpc.HTTPClient, cfg, v.BlockNumber, v.ValidatorPubkey)
+		if err != nil {
+			logger.Error("opt-in check failed", "slot", v.Slot, "error", err)
+			continue
+		}
+		if uerr := db.UpdateValidatorOptInStatus(ctx, v.Slot, opted); uerr != nil {
+			logger.Error("opt-in status update failed", "slot", v.Slot, "error", uerr)
+		}
+	}
+	if err := recentBids(ctx, db, httpc, relays, slotsForBids); err != nil {
```

Comment:
So again, I am not sure you understood what I meant when I asked you to use channels. The idea was that we do it one by one, so get missing slot, then get the validator info and opted in status for that slot and then get the bids for that slot, then move onto the next slot.

You have just changed this validator check to be sequential but then the bids again use batches. This does not make any sense. Either you do one by one or do batch like you were doing before.